pkg/fqdn/doc.go

```diff
@@ -12,20 +12,238 @@
+//    toPorts:
+//    - ports:
+//       - port: "53"
+//         protocol: ANY
+//      rules:
+//        dns:
+//          - matchPattern: "*"
+//          - matchName: "cilium.io"
+//
+// These redirects are implemented by the datapath and the management logic is
+// shared with other proxies in cilium (envoy and kafka). L7 DNS rules can
+// apply to an endpoint from various policies and, if any allow a request, it
+// will be forwarded to the original target of the DNS packet. This is often
+// configured in /etc/resolv.conf for a pod and k8s sets this automatically
+// (https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-dns-config)
+// In the example above `matchPattern: "*"` allows all requests and makes
+// `matchName: "cilium.io"` redundant.
+// Notes:
+//  - The forwarded requests are sent from cilium-agent on the host interface
+//    and not from the endpoint
+//  - Users must explicitly allow `*.*.svc.cluster.local.` in k8s clusters.
+//    This is not automatic
+//  - L7 DNS rules are egress-only
+//  - The proxy emits L7 cilium-monitor events: one for the request and a
+//    accept/reject event, and the final response
+//
+// Apart from allowing or denying DNS requests the DNS proxy is used to observe
+// DNS lookups in order to then allow L3 connections with the response
+// information. These must separately allowed with toFQDNs L3 rules. The
+// example above is a common "visibility" policy that allows all requests but
+// ensures that they traverse the proxy. This information is then placed in the
+// per-Endpoint and global DNS lookup caches and propagates from there.
+//
+// === L3 DNS ===
+// L3 DNS rules control L3 connections and not the DNS requests themselves.
+// They rely on DNS lookup cache information and it must come from the DNS
+// proxy, via a L7 DNS rule, or from the DNS poller (deprecated in cilium 1.8).
+//
+//  - toFQDNs:
+//      - matchName: "my-remote-service.com"
+//      - matchPattern: "bucket.*.my-remote-service.com"
+//
+// IPs seen in a DNS response (i.e. the request was allowed by a L7 policy)
+// that are also selected in a DNS L3 rule matchPattern or matchName have a /32
+// or /128 CIDR identity created. This occurs when they are first passed to the
+// toFQDN selectors from NameManager. These identities are not special in any
+// way and can overlap with toCIDR rules in policiies. They are placed in the
+// node-local ipcache and in the policy map of each endpoint that is allowed to
+// connect to them (i.e. defined in the L3 DNS rule).
+// Notes:
+// - Generally speaking, toFQDNs can only handle non-cluster IPs. In-cluster
+//   policy should use toEndpoints and toServices. This is partly historical but
+//   is because of ipcache limitations when mapping ip->identity. Endpoint
+//   identities can clobber the FQDN IP identity
+// - Despite being tracked per-Endpoint. DNS lookup IPs are collected into a
+//   global cache. This is historical and can be changed
```

Comment:
missing period.

Reply:
oh, this was an experiment on my part. I usually have proper periods in lists but I think I read somewhere that one isn't supposed to... so I tried it out. I actually did try to look it up and I think I'll add them back in to satisfy my OCD.